helix-term/src/commands.rs

```diff
@@ -2915,6 +2920,66 @@ fn goto_prev_diag(cx: &mut Context) {
     goto_pos(editor, pos);
 }
 
+fn goto_first_change(cx: &mut Context) {
+    goto_first_hunk_impl(cx, false);
+}
+
+fn goto_last_change(cx: &mut Context) {
+    goto_first_hunk_impl(cx, true);
+}
+
+fn goto_first_hunk_impl(cx: &mut Context, reverse: bool) {
+    let editor = &mut cx.editor;
+    let (_, doc) = current!(editor);
+    if let Some(handle) = doc.diff_handle() {
+        let hunk = {
+            let hunks = handle.hunks();
+            let idx = if reverse {
+                hunks.len().saturating_sub(1)
+            } else {
+                0
+            };
+            hunks.nth_hunk(idx)
+        };
+        if hunk != Hunk::NONE {
+            let pos = doc.text().line_to_char(hunk.after.start as usize);
+            goto_pos(editor, pos)
```

Comment:
Instead of selecting the point of start, I think these functions should select the whole hunk:

```rust
Selection::single(
  text.line_to_char(hunk.after.start),
  text.line_to_char(hunk.after.end),
)
```

like #4713

This lets you operate on the hunk once you jump to it (delete a hunk that's new or change some part of the hunk). Along these lines we may also want to add `mi`/`ma` textobjects so that you can select a hunk from within and operate on that.

Reply:
I think `mig` and `mag` would definitely be useful.

I have some reservations about making jump to hunk a select motion tough. Deletion hunks are special in that they really correspond to lines being removed BEFORE the line where the gutter is shown so they do not actually correspond to any text that is still in the document. They are shown at the first unchanged line after the removal (as bar at the top.to make that clear).

I think selecting this unchanged the line for deletions is inconsistent as it does not select any changed text. Ideally we really want a selection of width 0 at the star of the line. I don't believe that is possible tough, right?


An alternative could be to just select the first char of the unchanged line for deletions but that also feels inconsistent. 

Ideally we would habe separate motions for select changed lines and jump to hunk (the former being a selection and the latter being just a jump) but that would be yet anthor keybdinging (but potrntially a useful one) I am not sure what the best option would be . What are your thaughts here?

Reply:
I added a surround textobject (`mig`/`mag`) for now.
This surround object only matches the added text (so it doesn't do anything for a line where a removal is shown).

Reply:
Just giving my 2 cents...

I agree with @pascalkuthe that it feels more appropriate to not implement this as selection.

From a consistency standpoint, except _addition hunks_, you can't really accurately select what you _really_ changed anyways. So we might as well just choose the start position in all cases.

From a use case perspective, personally I do the hunk jumps when:

- I'm reviewing code - not taking action on the hunk itself
- I'm picking up my work-in-progress code from my last work session, and jumping helps me get there quickly. While I do need to take action on the code (continue working), it's unlikely I will have to action on the _whole_ hunk

In both cases, the hunk range can be easily identified through the gutters. We don't need the selection for that. And if it's unlikely that we need to act on the whole hunk, it seems unnecessary to select the thing by default.

After all, if you _really_ need to select it, you can do a `mig` afterwards.

Just my 2 cents from my experience using equivalent features on different editors (including this pending PR).

Reply:
> you can't really accurately select what you really changed anyways

Note yet ;) I am planning to add high quality word diffs (similar to delta or the github diff)  to `imara-diff` in the future. Then `mig` could select the words/char that were added and `mag`  could select the entire additions... That's a bit away tough

Reply:
> There is really no such thing as an outer hunk

I think this is ok, the same tends to happen for tree-sitter comment and parameter textobjects :+1:

> An alternative could be to just select the first char of the unchanged line for deletions but that also feels inconsistent.

I think this behavior would be ok. I don't think that jumping to deletion hunks would be very useful but also I think it would be surprising if the textobject totally disregarded them (skipped over them).

> Ideally we would habe separate motions for select changed lines and jump to hunk

I would be interested in exploring this idea where we have stable and consistent language for the different kinds of navigation (currently it's a bit muddled). Having extra commands that don't have keybinds is ok and then people could choose their desired behavior with just keymaps. For the default keybinds though I think it's good to be very selection-oriented: in testing this PR I found these textobjects to be very useful as ways to both jump to recently edited content (when a file is big and I don't know where I was exactly) and as a way to act on blocks that were modified/added (I didn't really think I would need this but it's actually very handy).

> In both cases, the hunk range can be easily identified through the gutters. We don't need the selection for that

Actually I disagree. Being able to see the range I would select in the gutter and then being able to easily select it is exactly the behavior I want from this textobject.

> After all, if you really need to select it, you can do a mig afterwards.

Conversely, and more consistently with other implemented textobjects, if you want the behavior where you select a point, you can bind to the textobject plus `;` (collapse_selection). Selecting as little as possible as default behavior directly fights against select-then-act IMO. Plus, using points for `]<textobject>`/`[<textobject>` makes it much much harder to extend in select mode.

Reply:
Yeah I'm actually indifferent - both sound reasonable to me, selection or not. I just feel like not selecting the thing is more consistent with the current diagnostics navigation `[d` and `]d`, which doesn't select the range.

Reply:
Yeah it would be 👍. I would like to change that behavior though - acting on the diagnostic range I think can be useful too (#4713)

Reply:
@the-mikedavis sorry I somehow missed the review email and won't be able to fix this today. I thaught about this some more and will change this to select the whole hunk.

I was just over thinking this. It might mean that a couple multicursor use cases don't work as expected but those cases also wouldn't work when we just jump to start (for those cases I would really want a separate motion to just select insertions but I will.leave that to the future)

Reply:
changed the motion the select the hunk